### ISSUE TITLE
Commit df583073e breaks zfs-auto-snapshot

### ISSUE THREAD

**Issue by jwittlincohen** [zfsonlinux/zfs#8591]
### System information

Type | Version/Name 
--- | --- 
Distribution Name	|  Debian
Distribution Version	|  Stretch (9.8)
Linux Kernel	|  4.9.0-8-amd64 #1 SMP Debian 4.9.144-3.1 (2019-02-19)
Architecture	|  x86_64
ZFS Version	| 0.8.0-rc3_141_gd93d4b1ac
SPL Version	| 0.8.0-rc3_141_gd93d4b1ac

### Describe the problem you're observing

Since building ZFS with commit df583073e, zfs-auto-snapshot no longer prunes old snapshots.  I think I know what's happening.  On prior version of master and 0.7, `zfs list -t snapshot` would list snapshots for all descendant filesystems.  As of commit df583073e, it only does so for the top-level dataset.  As a result, zfs-auto-snapshot doesn't see any snapshots to prune and continues creating new snapshots indefinitely.  Also, you can see that this problem started on my system around 4/1, the same day the commit was added to MASTER.

zfs-auto-snapshot uses the following command to find existing snapshots: `zfs list -H -t snapshot -o name -s name`

Here's what the output looks like on 0.7 and 0.8.0-rc3_141_gd93d4b1ac

0.7: https://pastebin.com/LhbYdmLM

0.8: 

root@backup-server:~# zfs list -H -t snapshot -o name -s name
bigbackup@zfs-auto-snap_daily-2019-02-21-0508
bigbackup@zfs-auto-snap_daily-2019-02-22-0508
bigbackup@zfs-auto-snap_daily-2019-02-23-0505
bigbackup@zfs-auto-snap_daily-2019-02-24-0508
bigbackup@zfs-auto-snap_daily-2019-02-25-0507
bigbackup@zfs-auto-snap_daily-2019-02-26-1527
bigbackup@zfs-auto-snap_frequent-2019-02-26-1530
bigbackup@zfs-auto-snap_frequent-2019-02-26-1545
bigbackup@zfs-auto-snap_frequent-2019-02-26-1600
bigbackup@zfs-auto-snap_frequent-2019-02-26-1615
bigbackup@zfs-auto-snap_hourly-2019-02-25-0517
bigbackup@zfs-auto-snap_hourly-2019-02-25-0617
bigbackup@zfs-auto-snap_hourly-2019-02-25-0717
bigbackup@zfs-auto-snap_hourly-2019-02-25-0817
bigbackup@zfs-auto-snap_hourly-2019-02-25-0917
bigbackup@zfs-auto-snap_hourly-2019-02-25-1017
bigbackup@zfs-auto-snap_hourly-2019-02-25-1117
bigbackup@zfs-auto-snap_hourly-2019-02-25-1217
bigbackup@zfs-auto-snap_hourly-2019-02-25-1317
bigbackup@zfs-auto-snap_hourly-2019-02-25-1417
bigbackup@zfs-auto-snap_hourly-2019-02-25-1517
bigbackup@zfs-auto-snap_hourly-2019-02-25-1617
bigbackup@zfs-auto-snap_hourly-2019-02-25-1717
bigbackup@zfs-auto-snap_hourly-2019-02-25-1817
bigbackup@zfs-auto-snap_hourly-2019-02-25-1917
bigbackup@zfs-auto-snap_hourly-2019-02-25-2017
bigbackup@zfs-auto-snap_hourly-2019-02-25-2117
bigbackup@zfs-auto-snap_hourly-2019-02-25-2217
bigbackup@zfs-auto-snap_hourly-2019-02-25-2317
bigbackup@zfs-auto-snap_hourly-2019-02-26-0017
bigbackup@zfs-auto-snap_hourly-2019-02-26-0117
bigbackup@zfs-auto-snap_hourly-2019-02-26-0217
bigbackup@zfs-auto-snap_hourly-2019-02-26-0317
bigbackup@zfs-auto-snap_hourly-2019-02-26-1617
bigbackup@zfs-auto-snap_weekly-2019-02-25-0512
rpool-backup@2-19-19
rpool-backup@2-20-19
rpool-backup@3-28-17
rpool-backup@3-29-17
rpool-backup@4-10-17
rpool-backup@4-3-17
rpool-backup@BACKUP-2-25-19
rpool-backup@__simplesnap_mainset_2018-09-21T00:50:02__
rpool-backup@__simplesnap_mainset_2018-11-05T12:43:21__
rpool-backup@__simplesnap_mainset_2019-02-25T21:23:33__
rpool-backup@__simplesnap_mainset_2019-03-01T10:58:38__
rpool-backup@__simplesnap_rpool2_2019-04-07T03:01:01__
rpool-backup@openssh-6.7p1
rpool-backup@pre-0.8.0

The 0.8 output is only showing the top level-directory.  For example, here's with -d4

https://pastebin.com/jJvzwFq8

### Describe how to reproduce the problem

Install a recent master with commit df583073e.  Install zfs-auto-snapshot, enable frequent snapshots, wait for it to generate more than 4 snapshots, and watch as zfs-auto-snapshot continues to generate new snapshots without ever pruning the old ones, regardless of the setting specified in /etc/cron.d/zfs-auto-snapshot. 

`*/15 * * * * root /usr/local/sbin/zfs-auto-snapshot -q -g --label=frequent --keep=4  //
`

```
Apr  7 09:15:02 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1315, 7 created, 0 destroyed, 0 warnings.
Apr  7 09:30:01 backup-server CRON[19859]: (root) CMD (zfs-auto-snapshot --quiet --syslog --label=frequent --keep=4  //)
Apr  7 09:30:02 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1330, 7 created, 0 destroyed, 0 warnings.
Apr  7 09:45:01 backup-server CRON[8895]: (root) CMD (zfs-auto-snapshot --quiet --syslog --label=frequent --keep=4  //)
Apr  7 09:45:02 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1345, 7 created, 0 destroyed, 0 warnings.
Apr  7 10:00:01 backup-server CRON[30264]: (root) CMD (zfs-auto-snapshot --quiet --syslog --label=frequent --keep=4  //)
Apr  7 10:00:02 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1400, 7 created, 0 destroyed, 0 warnings.
Apr  7 10:15:01 backup-server CRON[19381]: (root) CMD (zfs-auto-snapshot --quiet --syslog --label=frequent --keep=4  //)
Apr  7 10:15:02 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1415, 7 created, 0 destroyed, 0 warnings.
Apr  7 10:30:01 backup-server CRON[10195]: (root) CMD (zfs-auto-snapshot --quiet --syslog --label=frequent --keep=4  //)
Apr  7 10:30:01 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1430, 7 created, 0 destroyed, 0 warnings.
```

To resolve the issue, I added a `-d5` to each `zfs list -t snapshot` invoked by zfs-auto-snapshot

Now, it prunes old snapshots as expected:

`Apr  7 12:02:47 backup-server zfs-auto-snap: @zfs-auto-snap_frequent-2019-04-07-1600, 4 created, 1087 destroyed, 0 warnings.
`

It also works now when run manually:

```
root@backup-server:~# zfs-auto-snapshot --label=frequent --keep=4 --destroy-only rpool-backup/ROOT/debian
@zfs-auto-snap_frequent-2019-04-07-1612, 0 created, 548 destroyed, 0 warnings.

```

**Comment by bmaxa** [comment]
Same issue as mine. So this is not bug, but change of earlier behaviour, which can break anything
that relies on zfs list -t snapshot...

**Comment by jwittlincohen** [comment]
@bmaxa Yeah, any script that relies on the old behavior will have to be modified.

**Comment by behlendorf** [comment]
> I'm not sure if it was intended for zfs list -t snapshot

Thanks for catching this!  This was an inadvertent change in behavior, we definitely didn't intend to change the `zfs list -t snapshot` output.  Only to additionally support `zfs list -t snapshot <dataset>`.  I've opened #8593 with the proposed fix which resolves the change in behavior.  However, it would be great if someone experiencing this issue could verify it as well.

**Comment by jwittlincohen** [comment]
@behlendorf Thanks for the quick fix! I'm happy to rebuild ZFS with the included #8594 patch to test whether it resolves the issue.
